com.unity.multiplayer.mlapi/Tests/Runtime/NetworkUpdateLoopTests.cs

```diff
@@ -11,8 +11,45 @@ namespace MLAPI.RuntimeTests
 {
     public class NetworkUpdateLoopTests
     {
+        [UnityTest]
+        public IEnumerator RegisterAndUnregisterSystems()
+        {
+            // caching the current PlayerLoop (it will have NetworkUpdateLoop systems registered)
+            var cachedPlayerLoop = PlayerLoop.GetCurrentPlayerLoop();
+            {
+                // since current PlayerLoop already took NetworkUpdateLoop systems inside,
+                // we are going to swap it with the default PlayerLoop temporarily for testing
+                PlayerLoop.SetPlayerLoop(PlayerLoop.GetDefaultPlayerLoop());
+                var oldPlayerLoop = PlayerLoop.GetCurrentPlayerLoop();
+
+                NetworkUpdateLoop.RegisterLoopSystems();
+
+                int waitFrameNumber = Time.frameCount + 8;
+                yield return new WaitUntil(() => Time.frameCount >= waitFrameNumber);
+
+                NetworkUpdateLoop.UnregisterLoopSystems();
+
+                var newPlayerLoop = PlayerLoop.GetCurrentPlayerLoop();
+
+                // recursively compare old and new PlayerLoop systems and their subsystems
+                AssertAreEqualPlayerLoopSystems(newPlayerLoop, oldPlayerLoop);
+            }
+            // replace the current PlayerLoop with the cached PlayerLoop after the test
+            PlayerLoop.SetPlayerLoop(cachedPlayerLoop);
+        }
+
```

Comment:
```suggestion

        [Test]
        public void RegisterLoopSystemsShouldPreserveCustomizations()
        {
            PlayerLoop.SetPlayerLoop(PlayerLoop.GetDefaultPlayerLoop());
            NetworkUpdateLoop.RegisterLoopSystems();

            var loop = PlayerLoop.GetCurrentPlayerLoop();
            int subsystemCount = loop.subSystemList[0].subSystemList.Length + 1;

            var newSubsystemList = new PlayerLoopSystem[subsystemCount];
            Array.Copy(loop.subSystemList[0].subSystemList, newSubsystemList, loop.subSystemList[0].subSystemList.Length);
            newSubsystemList[newSubsystemList.Length - 1] = new PlayerLoopSystem
            {
                type = typeof(NetworkUpdateLoopTests),
            };
            loop.subSystemList[0].subSystemList = newSubsystemList;

            NetworkUpdateLoop.UnregisterLoopSystems();

            var finalLoop = PlayerLoop.GetCurrentPlayerLoop();

            // -2 because length - 1, and then another 1 for the networkinitialize having been removed.
            Assert.AreEqual(subsystemCount, finalLoop.subSystemList[0].subSystemList.Length - 2);
            Assert.AreEqual(typeof(NetworkUpdateLoopTests), finalLoop.subSystemList[0].subSystemList[subsystemCount - 2]);
        }
```

This test fails... I don't think it should (but please double-check as I was rushed before a meeting!)

Reply:
The test is indeed incorrect - it's missing registration of `loop` on or around line 57... still fails with that fix though.

Reply:
yeah, there are some errors with it, I think everything works just fine.

fix #1: set the new PlayerLoop after we add our new system there
```
loop.subSystemList[0].subSystemList = newSubsystemList;
PlayerLoop.SetPlayerLoop(loop); // <- this line was missing
```

fix #2: `Length - 2` logic is incorrect, it should be +1 because `NetworkUpdateLoop.UnregisterLoopSystems()` would only unregister `NetworkInitialization` system, not the one we injected. it should be `Length + 1` not `Length - 2`.
```
Assert.AreEqual(subsystemCount, finalLoop.subSystemList[0].subSystemList.Length - 2);
// to
Assert.AreEqual(subsystemCount, finalLoop.subSystemList[0].subSystemList.Length + 1);
// or
Assert.AreEqual(subsystemCount - 1, finalLoop.subSystemList[0].subSystemList.Length);
```

fix #3: we should compare **.type** property of a `PlayerLoopSystem`, not the type of the `PlayerLoopSystem` itself
```
Assert.AreEqual(typeof(NetworkUpdateLoopTests), finalLoop.subSystemList[0].subSystemList[subsystemCount - 2]);
// to
Assert.AreEqual(typeof(NetworkUpdateLoopTests), finalLoop.subSystemList[0].subSystemList[subsystemCount - 2].type);
```